x-pack/plugins/alerting/server/rules_client/common/apply_bulk_edit_operation.test.ts

```diff
@@ -210,16 +280,16 @@ describe('applyBulkEditOperation', () => {
       const ruleMock = {
         actions: [{ id: 'mock-action-id', group: 'default', params: {} }],
       };
-      expect(
-        applyBulkEditOperation(
-          {
-            field: 'schedule',
-            value: { interval: '1d' },
-            operation: 'set',
-          },
-          ruleMock
-        )
-      ).toHaveProperty('schedule', { interval: '1d' });
+      const { modifiedAttributes, isAttributeModified } = applyBulkEditOperation(
+        {
+          field: 'schedule',
+          value: { interval: '1d' },
+          operation: 'set',
+        },
+        ruleMock
+      );
+      expect(modifiedAttributes).toHaveProperty('schedule', { interval: '1d' });
+      expect(isAttributeModified).toBe(true);
```

Comment:
Quick question with regards to not seeing the `skipped` behavior when updating schedules. E.g. if you update the schedule to the same values twice, it doesn't say any rules were skipped for the second action:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/207964777-3ec9997c-fc52-4ee6-a4e9-beaece4f514b.gif" />
</p>

Is this because the `set_schedule` BulkActionEditType [is converted](https://github.com/elastic/kibana/blob/21de750b3cff4d4ad27d07c77135327e644b3bee/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/action_to_rules_client_operation.ts#L90-L100) to a `set` operation, and so it always overrides (as also mentioned in the UI warning callout)?

Debugging a bit shows `isAttributesUpdateSkipped` is `true` as the interval value hadn't changed (`"interval": "5m"`), however it looks like `isParamsUpdateSkipped` is `false`, presumably because this operation needed to be broken out to separate param updates (`to`, `from` and `interval`)?


Just curious here as I was hoping to piggy-back off your skipped logic here for deciding when `revision` should be updated over in https://github.com/elastic/kibana/pull/147398. Would appreciate any additional details here -- thanks! 🙂

Reply:
Seeing similar behavior for `Rule Actions` and `Timeline Templates`. This makes sense for `Rule Actions` since it's either an append or overwrite (so always updates), but why for timeline templates? Is this because it results in similar param updates as well? i.e.

https://github.com/elastic/kibana/blob/3287afd667b3ed42f848c87bd5faff1bc8d1ba80/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/rule_params_modifier.ts#L142-L151

Reply:
Maybe this is caused by the fact that currently we always increment the `params.version`?..

Reply:
So I think you have a point here @spong. The issue you are seeing is that we are not currently not checking in the case of `set_timeline` or `set_schedule` if the update should be skipped (like we do for index patterns, for e.g.) :https://github.com/elastic/kibana/blob/f9fb1441db5b4fb323aefb84db703d70a19c96dd/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/rule_params_modifier.ts#LL143C2-L70C2

I think we should add two checks `shouldSkipTimelineBulkAction` and `shouldSkipScheduleBulkAction` which should check if the new value is the same as the existing one saved in params. (For the case of Schedule, it is true that the update is split between `params` and `attributes`, but as you said the attributes checks already returns `true` because it is a generic check for all updated; so adding the check in the `params` will make the whole update to be marked as `skipped`).

The case of `actions` I still think needs a little more discussion. My initial thought was that if an action is:
- `add`ed: the result will always be different to the existing one, so rule is modified.
- `set`: even in the case of overwriting all actions with an action, that action is new and different to the existing one, so it also modifies the rule. My doubt now is: that is true for users using the UI, but we might have the case of a user using the API overwriting the actions to be exactly what they currently have? In that case, we'd need to mark it as skipped.

What do yo think?

cc @banderror

Reply:
Also would preferably do it in a follow up PR, if that doesn't disturb your workflow @spong . So we can get this one merged.

Reply:
Thank you for the details @jpdjere! This all makes sense now 🙂

I'm +1 for adding two additional checks for skipping `timeline` and `schedule` updates based on `params` (and thus aligning with the checks against `attributes`). 

As for `actions`, I understand your doubt, and so it would probably make sense to do a skip check for the sake of API consistency. That said, this may get a little interesting with how the SO references array gets converted over to the rule actions object. The `id` ends up being the connector id (the actual action SO id isn't leaked), so you can end up with what looks like two 'identical' actions on the rule itself:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/208173382-355f39c2-d120-4450-b504-47adb8f9696c.png" />
</p>


The rule update API handles this based on the array order IIRC....looking further that all appears to be isolated to [extractReferences()](https://github.com/elastic/kibana/blob/a3220fe1b6f02f1d5fdd70e0f386cc19058e3f97/x-pack/plugins/alerting/server/rules_client/lib/extract_references.ts#L16) and [denormalizeActions()](https://github.com/elastic/kibana/blob/a3220fe1b6f02f1d5fdd70e0f386cc19058e3f97/x-pack/plugins/alerting/server/rules_client/lib/denormalize_actions.ts#L14), so I might be over complicating this, and we can just do the skip diff once the actions are denormalized.

---

As for timing, I say let's get this PR in and we can tackle the above in a follow-up PR. It's no issue for me either way (I'm rebasing off this branch for the time being), but appreciate the thoughtfulness!